calcom/cal.com
# fix: identify org-wide features per user

## What does this PR do?

This pull request includes significant changes to the feature flag system, particularly focusing on the refactoring of how user feature checks are performed. The most important changes include the introduction of a new controller for checking user features, the removal of team-related feature checks, and the addition of comprehensive tests for the new implementations.

Refactoring feature flag checks:

* [`packages/emails/templates/organizer-request-email.ts`](diffhunk://#diff-3c3494160e97aa697f6dd86799bb7199935fee71160473be07f93abce1a70e14L1-R1): Replaced `FeaturesRepository` with `checkIfUserHasFeatureController` to determine if a user has access to a specific feature. Simplified the `getOrganizerRequestTemplate` function to only require `userId`. [[1]](diffhunk://#diff-3c3494160e97aa697f6dd86799bb7199935fee71160473be07f93abce1a70e14L1-R1) [[2]](diffhunk://#diff-3c3494160e97aa697f6dd86799bb7199935fee71160473be07f93abce1a70e14L10-R18)

* [`packages/features/flags/features.repository.interface.ts`](diffhunk://#diff-79f741e4197c18ef831e13f9f0e8785f106a0f3626ee68d4d5babac72ba6293eL6-L7): Removed methods related to team feature checks (`checkIfTeamHasFeature`, `checkIfTeamOrUserHasFeature`).

* [`packages/features/flags/features.repository.ts`](diffhunk://#diff-b7879e2cf35124244e4a99a57e76542abec730aa55d673b4eae100510eea227cL18-R55): Refactored `checkIfUserHasFeature` to include a check for whether a user belongs to a team with the feature. Added a private method `checkIfUserBelongsToTeamWithFeature` to encapsulate this logic.

Introduction of new controller:

* [`packages/features/flags/operations/check-if-user-has-feature.controller.ts`](diffhunk://#diff-96ab34adea85077faa1ea2f574b934fca59322f84356806449a9798646755ab6R1-R32): Introduced `checkIfUserHasFeatureController` to handle user feature checks, including input validation and orchestration of the use case.

Addition of tests:

* [`packages/features/flags/operations/check-if-user-has-feature.controller.test.ts`](diffhunk://#diff-5f0a3ee698056585fe7689b470cad76ba4cf326820906a4459be938fa5380403R1-R13): Added tests for `checkIfUserHasFeatureController` to ensure it correctly determines user feature access.
* [`packages/features/flags/operations/check-if-user-has-feature.use-case.test.ts`](diffhunk://#diff-919981605318ad5c12e9f171c095eaade5f70009b896d195c1df341003da7e26R1-R11): Added tests for `checkIfUserHasFeatureUseCase` to verify its functionality.

## Mandatory Tasks (DO NOT REMOVE)


## How should this be tested?

1. Ensure all existing feature flag checks still work correctly throughout the application.
2. Test the `checkIfUserHasFeature` function with various user IDs and feature slugs.
3. Verify that the DI container is properly initialized in different environments (development, test, production).
4. Run the new tests in `check-if-user-has-feature.controller.test.ts` and `check-if-user-has-feature.use-case.test.ts`.

## Checklist

- I have read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code follows the style guidelines of this project
- I have commented my code, particularly in hard-to-understand areas
- I have checked if my changes generate no new warnings
